dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/notification/DefaultTrackerNotificationWebHookService.java

```diff
@@ -0,0 +1,157 @@
+
+    @Override
+    @Transactional
+    public void handleEnrollment( String pi )
+    {
+        ProgramInstance instance = programInstanceStore.getByUid( pi );
+
+        List<ProgramNotificationTemplate> templates = instance.getProgram().getNotificationTemplates()
+            .stream()
+            .filter( t -> ProgramNotificationRecipient.WEB_HOOK == t.getNotificationRecipient() )
+            .collect( Collectors.toList() );
+
+        Map<String, String> payload = new HashMap<>();
+        ProgramNotificationMessageRenderer.VARIABLE_RESOLVERS
+            .forEach( ( key, value ) -> payload.put( key.name(), value.apply( instance ) ) );
+        sendPost( templates, renderService.toJsonAsString( payload ) );
+    }
+
+    @Override
+    @Transactional
+    public void handleEvent( String psi )
+    {
+        ProgramStageInstance instance = programStageInstanceStore.getByUid( psi );
+
+        List<ProgramNotificationTemplate> templates = instance.getProgramStage().getNotificationTemplates()
+            .stream()
+            .filter( t -> t.getNotificationRecipient() == ProgramNotificationRecipient.WEB_HOOK )
+            .collect( Collectors.toList() );
+
+        Map<String, String> payload = new HashMap<>();
+        ProgramStageNotificationMessageRenderer.VARIABLE_RESOLVERS
+            .forEach( ( key, value ) -> payload.put( key.name(), value.apply( instance ) ) );
+        sendPost( templates, renderService.toJsonAsString( payload ) );
+    }
+
+    private void sendPost( List<ProgramNotificationTemplate> templates, String payload )
+    {
+        ResponseEntity<String> responseEntity = null;
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.put( "Content-type", Lists.newArrayList( "application/json" ) );
+
+        HttpEntity<String> httpEntity = new HttpEntity<>( payload, httpHeaders );
+
+        for ( ProgramNotificationTemplate t : templates )
+        {
+            URI uri = UriComponentsBuilder.fromHttpUrl( t.getMessageTemplate() ).build().encode().toUri();
+
+            try
+            {
+                responseEntity = restTemplate.exchange( uri, HttpMethod.POST, httpEntity, String.class );
+            }
+            catch ( HttpClientErrorException ex )
+            {
+                log.error( "Client error " + ex.getMessage() );
+            }
+            catch ( HttpServerErrorException ex )
+            {
+                log.error( "Server error " + ex.getMessage() );
+            }
+            catch ( Exception ex )
+            {
+                log.error( "Error " + ex.getMessage() );
+            }
+
+            if ( responseEntity != null && SmsGateway.OK_CODES.contains( responseEntity.getStatusCode() ) )
+            {
+                log.info( "Post request to url: " + t.getMessageTemplate() + " successful." );
+            }
+            else
+            {
+                log.error( "Post request to url: " + t.getMessageTemplate() + " failed." );
```

Comment:
Maybe include the program or template uid? so its easier to trace and see where the error happened

Reply:
resolved